### ISSUE TITLE
Does not recognise custom html elements

### ISSUE THREAD

**Issue by timoxley** [chjj/marked#283]
### tldr;

[Custom html elements](www.html5rocks.com/en/tutorials/webcomponents/customelements/) are coming, and `marked` does not currently recognise them as valid html. Custom elements may have different escaping requirements for their content based on whether the elements contains more elements (like `div`) or plain text (like `script`). Perhaps it would be possible to pass which html tags to escape/not escape as configuration properties.
### Backstory

I've been experimenting with custom elements and I wanted to inject some into [my markdown presentation](https://github.com/timoxley/shadowdom-talk-slides).

`````` markdown
## Shadow DOM is a sub-tree inside a DOM node
#### Like a mini-iframe that replaces an Element.

<h3>Light DOM</h3>

```html
<h3>Light DOM</h3>
`` `
<play-script>
var heading = currentSlide.querySelector('h3')
var root = heading.createShadowRoot()
root.innerHTML = '<a href="#">Shadow DOM</a>'
</play-script>
``````

[Source](https://github.com/timoxley/shadowdom-talk-slides/blame/b260937e10764f363aba28c386cc40ad8f93c9ba/slides.md#L19-L31)

[`play-script`](https://github.com/timoxley/shadowdom-talk-slides/blob/b260937e10764f363aba28c386cc40ad8f93c9ba/client.js#L95-L111) is essentially a `script` tag that renders its content, highlighted with [highlight.js](https://github.com/isagalaev/highlight.js), and executes only when you click on it.

Problem: **currently `marked` does not recognise tags with hyphens in them as valid html**. To get around this quickly, I loosened the html detection regex to also match tags with a single hypen:

[Source](https://github.com/timoxley/marked/compare/feature;custom-html-elements#diff-81ab0a5aa39b7a91951fc28d2d151496L44)

``` diff
-  + '|span|br|wbr|ins|del|img)\\b)\\w+(?!:/|@)\\b';
+  + '|span|br|wbr|ins|del|img)\\b)[\\w-]+(?!:/|@)\\b';
```

This successfully recognised my custom tag `play-script` as html, but it raised a new issue:

My custom tag's content was text as opposed to more xml tags, so **marked would html-escape all of my script contents**. This was no good since I wanted that text to remain as executable script (perhaps I should have looked at unescaping).

So as a hack to make it work, I reluctantly added my specific tag, (and the `template` tag) as special-cased to avoid escaping, along with `style`, `pre` and `script`:

``` diff
-        pre: cap[1] === 'pre' || cap[1] === 'script' || cap[1] === 'style',
+        pre: cap[1] === 'pre' || cap[1] === 'script' || cap[1] === 'style' || cap[1] === 'play-script' || cap[1] === 'template',
```

[Source](https://github.com/timoxley/marked/compare/feature;custom-html-elements#diff-81ab0a5aa39b7a91951fc28d2d151496L44)

Special-casing my custom element is clearly not at all a solution, but perhaps the tags that are matched here could be passed in as configuration properties.

**Comment by lkiarest** [comment]
+1

**Comment by joshbruce** [comment]
#985
